<font color="#0D9DF0"><h3><B>Trabalho Final - Python para Ciência de Dados</b></h3></font>
***

# 

---
## Análise do Desenvolvimento Humano e Econômico Global com base nos dados obtidos no site Gapminder

#### Membros
1. Andre Augusto Ferreira Cabelo
2. Bruno Braga Guedes Cardoso
3. Bruno dos Santos Crivelli
4. Marcelo Tranche de Souza Junior

O objetivo da análise é averiguar o quanto índices de corrupção e guerras, por exemplo, resultam em um baixo desenvolvimento humano e qualidade de vida nos países

A base em questão reune métricas desde 1952 de fatores como renda per capta, expectativa de vida, continente e população. Além disso, foram consideradas outras 5 bases, também do Gapminder, que tem as seguintes nomenclaturas:

1. GDP Per Data (Price and inflation adjusted) - Contém os dados de GPD ajustado pela inflação, dado que a base original trás apenas os dados do ano corrente.
2. Extreme poverty rate (less then $2.15 ...) - Contém os dados do percentual da população que está na extrema pobreza, isto é, recebendo menos que $2,15 por dia.
3. Child mortality (per 1000 live births) - Contém a quantidade de crianças que morrem por ano a cada 1000 nascimentos
4. Human Development Index (HDI) - Contém o score do IDH de cada país por ano
5. Population Density (per square km) - Contém a densidade populacional por metro quadrado de cada país por ano
6. Mortes por conflitos armados - Contém o volume de pessoas que morreram, por ano, em conflitos armados

---
## 2. Coleta, Limpeza e Análise Exploratória (AED)

### Objetivo
Preparar os dados, tratando inconsistências e identificando padrões iniciais.

### Itens a incluir
- **Carregamento do dataset** Gapminder.  
- **Verificação de valores ausentes**.  
- **Remoção de duplicados**.  
- **Análise estatística descritiva inicial** (médias, medianas, desvios etc.).  
- **Visualizações exploratórias básicas** (ex: histogramas, boxplots, contagens por continente).  

### Markdown deve conter
- Justificativas das decisões tomadas.  
- Observações sobre a qualidade dos dados.  
- Interpretação dos primeiros resultados.

### Carregamento do Dataset

A primeira coisa a ser feita é importar os arquivos de classe auxiliares e fazer a leitura da base original do gapminder:

In [1]:
import sys
sys.path.append('../avaliacao-python-ciencia-dados/')

from utils.data_analisys import *
from utils.data_manipulation import *
from utils.database_manipulation import *

**_Momento setup de projeto_**

_O projeto encontra-se estruturado em alguns blocos, mas daremos um breve resumo da estrutura:_
- _Pasta Bases: Contém todas as bases utilizadas durante esse projeto. As bases originais estão disponíveis na raiz, enquanto as bases tratadas TODAS foram salvas na pasta **tratada**, isso inclui não só a tabela principal com a junção de todas as outras bases, mas também suas respectivas bases tratadas de forma individual._
- _Pasta Enunciado: Contém o enunciado do trabalho._
- _Pasta scr: Contém os códigos fonte do projeto. O arquivo notebook está disponível na raiz, mas também há uma pasta **utils** que contém as funções utilizadas ao longo de todo o projeto._

Em seguida, vamos importar a tabela principal e imprimir suas primeiras linhas, que está salva na pasta de Bases

In [2]:
base_gapminder = importa_csv(dataframe_name='../../avaliacao-python-ciencia-dados/Bases/gapminder_full.csv', separador=',', encoding='utf-8')
base_gapminder.head()

Importando o dataframe: ../../avaliacao-python-ciencia-dados/Bases/gapminder_full.csv ...
Dataframe importado com sucesso! 
Dimensão: (1736, 6)


country  year         pop continent  lifeExp   gdpPercap
0  Afghanistan  1952   8425333.0      Asia   28.801  779.445314
1  Afghanistan  1957   9240934.0      Asia   30.332  820.853030
2  Afghanistan  1962  10267083.0      Asia   31.997  853.100710
3  Afghanistan  1967  11537966.0      Asia   34.020  836.197138
4  Afghanistan  1972  13079460.0      Asia   36.088  739.981106

Em seguida, rodaremos 2 funções de análise e 1 de limpeza:
- Analisa Dados Ausentes: Lista quais registros tem colunas com dados faltantes
- Verifica Dados Duplicados: Lista quais registros são duplicados
- Remove Dados Duplicados: Remove as duplicatas da base, evitando cacofonia dos dados

In [3]:
analisa_dados_ausentes(base_gapminder)

Empty DataFrame
Columns: [variavel_analisada, Quantidade, Percentual]
Index: []

Rodando essa função, foi possível ver que não há dados ausentes na base.

In [4]:
verifica_dados_duplicados(base_gapminder)

Foram encontrados 32 registros duplicados

Prévia dos dados duplicados:


country  year         pop continent  lifeExp    gdpPercap
72   Afghanistan  1997  22227415.0      Asia   41.763   635.341351
145      Bolivia  1987   6156369.0  Americas   57.251  2753.691490
146      Bolivia  1992   6893451.0  Americas   59.957  2961.699694
147      Bolivia  1997   7693188.0  Americas   62.050  3326.143191
148      Bolivia  2002   8445134.0  Americas   63.883  3413.262690

Já esta demonstra que temos 32 duplicatas na base, que serão removidas a seguir:

In [5]:
base_limpa = remove_dados_duplicados(base_gapminder)
base_limpa.head()

Removendo dados duplicados ...
Foram encontrados 32 registros duplicados

Prévia dos dados duplicados:
Dados duplicados removidos com sucesso! 
Dimensão: (1704, 6)


country  year         pop continent  lifeExp   gdpPercap
0  Afghanistan  1952   8425333.0      Asia   28.801  779.445314
1  Afghanistan  1957   9240934.0      Asia   30.332  820.853030
2  Afghanistan  1962  10267083.0      Asia   31.997  853.100710
3  Afghanistan  1967  11537966.0      Asia   34.020  836.197138
4  Afghanistan  1972  13079460.0      Asia   36.088  739.981106

O arquivo continha 1736 linhas e agora possui apenas 1704, o que demonstra que não há mais duplicatas no arquivo.

Após isso, rodamos 2 comandos pra analisar a qualidade e formato dos dados:

In [6]:
base_limpa.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1704 entries, 0 to 1735
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   country    1704 non-null   object 
 1   year       1704 non-null   int64  
 2   pop        1704 non-null   float64
 3   continent  1704 non-null   object 
 4   lifeExp    1704 non-null   float64
 5   gdpPercap  1704 non-null   float64
dtypes: float64(3), int64(1), object(2)
memory usage: 93.2+ KB


In [7]:
base_limpa.describe()

year           pop      lifeExp      gdpPercap
count  1704.00000  1.704000e+03  1704.000000    1704.000000
mean   1979.50000  2.960121e+07    59.474439    7215.327081
std      17.26533  1.061579e+08    12.917107    9857.454543
min    1952.00000  6.001100e+04    23.599000     241.165876
25%    1965.75000  2.793664e+06    48.198000    1202.060309
50%    1979.50000  7.023596e+06    60.712500    3531.846988
75%    1993.25000  1.958522e+07    70.845500    9325.462346
max    2007.00000  1.318683e+09    82.603000  113523.132900

Nessa primeira análise identificamos que os tipos estão fazendo sentido de acordo com os valores registrados em cada coluna, apesar de ser estranho população estar em ponto flutuante. Ao analisar, descobrimos o seguinte:
#### Problemas encontrados olhando o dataset e as infos:
1. Índia -> Pop com Notação Científica - Verificar se precisa tratar
2. China -> Pop com decimal
3. Reduzir renda per capita para 2 casas decimais
4. Reduzir expectativa de vida para 1 casa decimal
5. Mínimo de Expectativa de vida de 23.599 é  estranho (Spolier alert: Não é estranho não rs)
6. double check pra ver se tem gente com ano faltando

Vamos começar a tratas esses problemas em seguida:

In [8]:
# Reduzir renda per capita para 2 casas decimais
base_limpa['gdpPercap'] = base_limpa['gdpPercap'].round(2)

# Reduzir expectativa de vida para 1 casa decimal
base_limpa['lifeExp'] = base_limpa['lifeExp'].round(1)

# Tratar população da Índia e China aproveitando já converte todo mundo logo
base_limpa["pop"] = base_limpa["pop"].astype("int64")
base_limpa["pop"] = base_limpa["pop"].apply(lambda x: f"{x:,.0f}".replace(",", "."))

# Rwanda teve uma guerra civil em 1994, o que impactou a expectativa de vida do país. Isso foi visto pq paramos no meio da análise aqui 
# pra achar e vimos que teve guerra, o que nos levou a buscar a base de guerras. O valor de 23.599 é real e não um erro de base, e vem de Rwanda.

# Analisar quais paises tem anos faltando -> Não parece ter impacto
anos_por_pais = base_limpa.groupby("country").count()["year"].reset_index().sort_values(by="year", ascending=True)
anos_por_pais.head(1000)


C:\Users\bruno\AppData\Local\Temp\ipykernel_30196\164088985.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  base_limpa['gdpPercap'] = base_limpa['gdpPercap'].round(2)
C:\Users\bruno\AppData\Local\Temp\ipykernel_30196\164088985.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  base_limpa['lifeExp'] = base_limpa['lifeExp'].round(1)
C:\Users\bruno\AppData\Local\Temp\ipykernel_30196\164088985.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[r

country  year
0           Afghanistan    12
1               Albania    12
2               Algeria    12
3                Angola    12
4             Argentina    12
..                  ...   ...
137             Vietnam    12
138  West Bank and Gaza    12
139          Yemen Rep.    12
140              Zambia    12
141            Zimbabwe    12

[142 rows x 2 columns]

Vamos tirar a teima pra China e India:

In [9]:
base_limpa[base_limpa['country'] == 'China'].head(30)

country  year            pop continent  lifeExp  gdpPercap
297   China  1952    556.263.527      Asia     44.0     400.45
298   China  1957    637.408.000      Asia     50.5     575.99
299   China  1962    665.770.000      Asia     44.5     487.67
300   China  1967    754.550.000      Asia     58.4     612.71
301   China  1972    862.030.000      Asia     63.1     676.90
302   China  1977    943.455.000      Asia     64.0     741.24
303   China  1982  1.000.281.000      Asia     65.5     962.42
304   China  1987  1.084.035.000      Asia     67.3    1378.90
305   China  1992  1.164.970.000      Asia     68.7    1655.78
306   China  1997  1.230.075.000      Asia     70.4    2289.23
307   China  2002  1.280.400.000      Asia     72.0    3119.28
308   China  2007  1.318.683.096      Asia     73.0    4959.11

In [10]:
base_limpa[base_limpa['country'] == 'India'].head(30)

country  year            pop continent  lifeExp  gdpPercap
728   India  1952    372.000.000      Asia     37.4     546.57
729   India  1957    409.000.000      Asia     40.2     590.06
730   India  1962    454.000.000      Asia     43.6     658.35
731   India  1967    506.000.000      Asia     47.2     700.77
732   India  1972    567.000.000      Asia     50.7     724.03
733   India  1977    634.000.000      Asia     54.2     813.34
734   India  1982    708.000.000      Asia     56.6     855.72
735   India  1987    788.000.000      Asia     58.6     976.51
736   India  1992    872.000.000      Asia     60.2    1164.41
737   India  1997    959.000.000      Asia     61.8    1458.82
738   India  2002  1.034.172.547      Asia     62.9    1746.77
739   India  2007  1.110.396.331      Asia     64.7    2452.21

Tudo certo, agora o arquivo parece pronto pra análise.

A seguir, começaremos a tratar todas as bases com as quais estamos lidando.

### Carregamento das bases Auxiliares

Começaremos com a base que mais deu trabalho, a base de guerras. As demais bases são semelhantes em estrutura, então o processo de tratamento é o mesmo, sendo este isolado num método apropriado.

A complexidade da base de mortes por guerra reside no fato de ser uma base descritiva anual, ou seja, eu tenho o número de mortes por ano, enquanto nosso dataset só possui registros de 5 em 5 anos. Além disso, o intervalo dessa base vai de 1989 à 2024, o que cria um período de "vácuo" onde essas bases não possuem dados. Isso será levado em conta na análise pra limitar os resultados desse indicador no período de 1989 a 2007, pois a base principal se encerra em 2007.

Dito isso, seguem os pontos necessários pra essa base:
1. Importar
2. Analisar dados ausentes e remover duplicadas se tiver
3. Agregar as mortes pros anos disponíveis na base principal, expurgando tudo que vier dps de 2007 e somando tudo que vier de 5 em 5 anos (Ex.: o dado de 2007 tem que ser 2003+2004+2005+2006+2007, pois óbitos é um indicador acumulativo, diferente dos demais dessa tabela)
4. Renomear as colunas principais do dataset, pois seus respectivos nomes são muito grandes pra serem devidamente analisados durante uma leitura ideal
5. Gerar uma coluna que some o total de mortes do período, pra uma análise consolidada
6. Por fim, agregar ao dataset principal pra gente ter tudo no mesmo lugar para futuras análises

O código abaixo tá comentado com cada passo:

In [11]:
# Tratar a base de mortes por guerra e adicionar na brincadeira:
mortes_guerra = importa_csv(dataframe_name='../../avaliacao-python-ciencia-dados/Bases/deaths-in-armed-conflicts-by-type.csv', separador=',', encoding='utf-8')

analisa_dados_ausentes(mortes_guerra)
verifica_dados_duplicados(mortes_guerra) # tem 0 registros duplicados

mortes_guerra.describe() #7632 registros


Importando o dataframe: ../../avaliacao-python-ciencia-dados/Bases/deaths-in-armed-conflicts-by-type.csv ...
Dataframe importado com sucesso! 
Dimensão: (7632, 8)
Foram encontrados 0 registros duplicados

Prévia dos dados duplicados:


Year  \
count  7632.000000   
mean   2006.500000   
std      10.388975   
min    1989.000000   
25%    1997.750000   
50%    2006.500000   
75%    2015.250000   
max    2024.000000   

       Deaths in ongoing conflicts (best estimate) - Conflict type: one-sided violence  \
count                                        7632.000000                                 
mean                                          465.031053                                 
std                                         15443.385459                                 
min                                             0.000000                                 
25%                                             0.000000                                 
50%                                             0.000000                                 
75%                                             0.000000                                 
max                                        779683.000000                                 

       Deaths in ongoing conflicts (best estimate) - Conflict type: non-state conflict  \
count                                        7632.000000                                 
mean                                          150.082154                                 
std                                          1198.977717                                 
min                                             0.000000                                 
25%                                             0.000000                                 
50%                                             0.000000                                 
75%                                             0.000000                                 
max                                         31054.000000                                 

       Deaths in ongoing conflicts (best estimate) - Conflict type: extrasystemic  \
count                                             7632.0                            
mean                                                 0.0                            
std                                                  0.0                            
min                                                  0.0                            
25%                                                  0.0                            
50%                                                  0.0                            
75%                                                  0.0                            
max                                                  0.0                            

       Deaths in ongoing conflicts (best estimate) - Conflict type: intrastate  \
count                                        7632.000000                         
mean                                          780.564465                         
std                                          6677.569455                         
min                                             0.000000                         
25%                                             0.000000                         
50%                                             0.000000                         
75%                                             0.000000                         
max                                        199567.000000                         

       Deaths in ongoing conflicts (best estimate) - Conflict type: interstate  
count                                        7632.000000                        
mean                                          149.083726                        
std                                          3112.485153                        
min                                             0.000000                        
25%                                             0.000000                        
50%                                             0.000000                        
75%                                             0.000000                        
max                                     

In [12]:
#Realizar agregação por pais e ano, equalizando a coluna de ano com as da base principal
#A base de mortes em guerra vai de 89 a 2024 e a base principal de 1952 a 2007. teremos apenas um pequeno intervalo de anos de interseção
# que será agregado e analisado
anos = [1952, 1957, 1962, 1967, 1972,1977,1982,1987,1992,1997,2002,2007]
colunas_agrupadas = ['Deaths in ongoing conflicts (best estimate) - Conflict type: one-sided violence','Deaths in ongoing conflicts (best estimate) - Conflict type: non-state conflict','Deaths in ongoing conflicts (best estimate) - Conflict type: extrasystemic','Deaths in ongoing conflicts (best estimate) - Conflict type: intrastate','Deaths in ongoing conflicts (best estimate) - Conflict type: interstate']
#renomeia coluna de ano pra ficar no padrão de programação e das demais tabelas (camelCase)
renomeia_coluna(mortes_guerra, nome_anterior='Year', novo_nome='year', inplace=True)

mortes_guerra = agrega_anos(mortes_guerra, anos, colunas_agrupadas,coluna_agrupamento = ['Entity'])
mortes_guerra.info() # 848 registros  -> 9 registros por país certinho

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 848 entries, 0 to 847
Data columns (total 7 columns):
 #   Column                                                                               Non-Null Count  Dtype 
---  ------                                                                               --------------  ----- 
 0   Entity                                                                               848 non-null    object
 1   Deaths in ongoing conflicts (best estimate) - Conflict type: one-sided violence_sum  848 non-null    int64 
 2   Deaths in ongoing conflicts (best estimate) - Conflict type: non-state conflict_sum  848 non-null    int64 
 3   Deaths in ongoing conflicts (best estimate) - Conflict type: extrasystemic_sum       848 non-null    int64 
 4   Deaths in ongoing conflicts (best estimate) - Conflict type: intrastate_sum          848 non-null    int64 
 5   Deaths in ongoing conflicts (best estimate) - Conflict type: interstate_sum          848 non-null   

In [13]:
# renomeiar colunas para facilitar o uso e entendimento do que tá rolando.
renomeia_coluna(mortes_guerra, nome_anterior='Entity', novo_nome='country', inplace=True)
renomeia_coluna(mortes_guerra, nome_anterior='Deaths in ongoing conflicts (best estimate) - Conflict type: one-sided violence_sum', novo_nome='oneSidedViolence', inplace=True)
renomeia_coluna(mortes_guerra, nome_anterior='Deaths in ongoing conflicts (best estimate) - Conflict type: non-state conflict_sum', novo_nome='nonStateConflict', inplace=True)
renomeia_coluna(mortes_guerra, nome_anterior='Deaths in ongoing conflicts (best estimate) - Conflict type: extrasystemic_sum', novo_nome='extrasystemic', inplace=True)
renomeia_coluna(mortes_guerra, nome_anterior='Deaths in ongoing conflicts (best estimate) - Conflict type: intrastate_sum', novo_nome='intrastate', inplace=True)
renomeia_coluna(mortes_guerra, nome_anterior='Deaths in ongoing conflicts (best estimate) - Conflict type: interstate_sum', novo_nome='interstate', inplace=True)

country  oneSidedViolence  nonStateConflict  extrasystemic  intrastate  \
0    Abkhazia                 0                 0              0           0   
1    Abkhazia                 0                 0              0           0   
2    Abkhazia                 0                 0              0           0   
3    Abkhazia                 0                 0              0           0   
4       Abyei                 0                 0              0           0   
..        ...               ...               ...            ...         ...   
843    Zambia                 0                 0              0           0   
844  Zimbabwe                51                 0              0           0   
845  Zimbabwe                 0                 0              0           0   
846  Zimbabwe                 0                 0              0           0   
847  Zimbabwe                 0                 0              0           0   

     interstate  year  
0             0  1992  
1             0  1997  
2             0  2002  
3             0  2007  
4             0  1992  
..          ...   ...  
843           0  2007  
844           0  1992  
845           0  1997  
846           0  2002  
847           0  2007  

[848 rows x 7 columns]

In [14]:
# Validando o caso de Rwanda pra ver se deu tudo certo
filtro = mortes_guerra['country'] == 'Rwanda'
ausentes = mortes_guerra[filtro].index
mortes_guerra.loc[ausentes].head()

# A soma pra Rwanda bateu certinho cruzando com a base original (fiz na calculadora mesmo rs)

country  oneSidedViolence  nonStateConflict  extrasystemic  intrastate  \
620  Rwanda              3013                 0              0        1366   
621  Rwanda            782261                92              0        3344   
622  Rwanda               708                 0              0        3139   
623  Rwanda                 0                 0              0           0   

     interstate  year  
620           0  1992  
621           0  1997  
622           0  2002  
623           0  2007

In [15]:
# Gerar uma coluna com o total de mortes por guerra
mortes_guerra['totalWarDeaths'] = mortes_guerra[['oneSidedViolence', 'nonStateConflict', 'extrasystemic', 'intrastate', 'interstate']].sum(axis=1)
mortes_guerra.loc[ausentes].head()

country  oneSidedViolence  nonStateConflict  extrasystemic  intrastate  \
620  Rwanda              3013                 0              0        1366   
621  Rwanda            782261                92              0        3344   
622  Rwanda               708                 0              0        3139   
623  Rwanda                 0                 0              0           0   

     interstate  year  totalWarDeaths  
620           0  1992            4379  
621           0  1997          785697  
622           0  2002            3847  
623           0  2007               0

In [16]:
#Agora a gente junta as 2 bases usando o left join pra garantir que todos os registros da original vão entrar, principalmente por sabermos que a
#base de guerras tem dados faltantes pra vários anos
juncao = junta_dataframes(base_limpa, mortes_guerra, chave=['country', 'year'], tipo_juncao='left')
juncao.head()

Dataframes unidos com sucesso! 
Dimensão: (1704, 12)


country  year         pop continent  lifeExp  gdpPercap  \
0  Afghanistan  1952   8.425.333      Asia     28.8     779.45   
1  Afghanistan  1957   9.240.934      Asia     30.3     820.85   
2  Afghanistan  1962  10.267.083      Asia     32.0     853.10   
3  Afghanistan  1967  11.537.966      Asia     34.0     836.20   
4  Afghanistan  1972  13.079.460      Asia     36.1     739.98   

   oneSidedViolence  nonStateConflict  extrasystemic  intrastate  interstate  \
0               NaN               NaN            NaN         NaN         NaN   
1               NaN               NaN            NaN         NaN         NaN   
2               NaN               NaN            NaN         NaN         NaN   
3               NaN               NaN            NaN         NaN         NaN   
4               NaN               NaN            NaN         NaN         NaN   

   totalWarDeaths  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

Agora seguiremos pras outras tabelas, cujo processo será sempre o mesmo:
1. Importar
2. Remover duplicadas se tiver
3. Transpor o dataset, pois ele tem uma coluna pra cada ano por pais, é necessário que ele tenha uma linha para cada país + ano para podermos cruzar com o dataset principal (Método desenvolvido com a ajuda de IA)
4. Renomear as colunas pra manter o padrão das bases
5. Arredondar os valores pra 2 casas decimais
6. Por fim, agregar ao dataset principal pra gente ter tudo no mesmo lugar para futuras análises

In [17]:
#Agora vamos agregar a tabela de GDP ajustado pela inflação
base_gdp_ajustado = importa_csv(dataframe_name='../../avaliacao-python-ciencia-dados/Bases/gdp_pcap_inflaction_adjusted.csv', separador=',', encoding='utf-8')
base_gdp_ajustado.head()

Importando o dataframe: ../../avaliacao-python-ciencia-dados/Bases/gdp_pcap_inflaction_adjusted.csv ...
Dataframe importado com sucesso! 
Dimensão: (193, 303)


geo         name        1800        1801        1802        1803  \
0  afg  Afghanistan   480.75476   480.75476   480.75476   480.75476   
1  ago       Angola   373.05144   374.25872   376.06965   377.88058   
2  alb      Albania   469.30821   470.65412   472.00389   473.35753   
3  and      Andorra  1370.15106  1372.44036  1374.72967  1378.16364   
4  are          UAE  1142.36487  1145.79884  1150.37745  1154.95607   

         1804        1805        1806        1807  ...         2091  \
0   480.75476   480.75476   480.75476   480.75476  ...   4684.97868   
1   379.08787   380.89880   382.70973   384.52066  ...  24542.76609   
2   474.71505   476.07647   477.44178   478.81102  ...  54455.30605   
3  1380.45294  1382.74225  1385.03156  1387.32087  ...  79877.72947   
4  1159.53469  1162.96865  1167.54727  1172.12588  ...  92569.33855   

          2092         2093         2094         2095         2096  \
0   4793.65905   4905.29816   5019.97187   5137.75735   5258.73302   
1  25049.27925  25561.99308  26080.79692  26605.57338  27136.19842   
2  55061.03873  55663.53627  56262.63285  56858.16757  57449.98463   
3  80165.94404  80449.12000  80727.31187  81000.57506  81268.96576   
4  92603.91018  92637.70452  92670.73861  92703.02914  92734.59243   

          2097         2098         2099         2100  
0   5382.97857   5510.57490   5641.60409   5776.14935  
1  27672.54146  28214.46550  28761.82726  29314.47731  
2  58037.93335  58621.86829  59201.64931  59777.14164  
3  81532.54087  81791.35793  82045.47508  82294.95092  
4  92765.44449  92795.60097  92825.07719  92853.88818  

[5 rows x 303 columns]

In [18]:
#Trata a tabela usando a função que vai ser reutilizada nas próximas tb
base_gdp_ajustado = transpoe_e_trata_dataframe(base_gdp_ajustado, nome_coluna_resultado='gdpPerCaptaInflactionAdjusted')
base_gdp_ajustado.head()

Removendo dados duplicados ...
Foram encontrados 0 registros duplicados

Prévia dos dados duplicados:
Dados duplicados removidos com sucesso! 
Dimensão: (58093, 3)


country  year  gdpPerCaptaInflactionAdjusted
0  Afghanistan  1800                         480.75
1  Afghanistan  1801                         480.75
2  Afghanistan  1802                         480.75
3  Afghanistan  1803                         480.75
4  Afghanistan  1804                         480.75

In [19]:
#Agora a gente junta as 2 bases pra ver no que dá
juncao = junta_dataframes(juncao, base_gdp_ajustado, chave=['country', 'year'], tipo_juncao='left')
juncao.head()

Dataframes unidos com sucesso! 
Dimensão: (1704, 13)


country  year         pop continent  lifeExp  gdpPercap  \
0  Afghanistan  1952   8.425.333      Asia     28.8     779.45   
1  Afghanistan  1957   9.240.934      Asia     30.3     820.85   
2  Afghanistan  1962  10.267.083      Asia     32.0     853.10   
3  Afghanistan  1967  11.537.966      Asia     34.0     836.20   
4  Afghanistan  1972  13.079.460      Asia     36.1     739.98   

   oneSidedViolence  nonStateConflict  extrasystemic  intrastate  interstate  \
0               NaN               NaN            NaN         NaN         NaN   
1               NaN               NaN            NaN         NaN         NaN   
2               NaN               NaN            NaN         NaN         NaN   
3               NaN               NaN            NaN         NaN         NaN   
4               NaN               NaN            NaN         NaN         NaN   

   totalWarDeaths  gdpPerCaptaInflactionAdjusted  
0             NaN                        1540.11  
1             NaN                        1708.81  
2             NaN                        1869.50  
3             NaN                        1930.54  
4             NaN                        1602.84

In [20]:
#Adicionando a base de extrema pobreza
base_extrema_pobreza = importa_csv(dataframe_name='../../avaliacao-python-ciencia-dados/Bases/gm_extreme_poverty_rate.csv', separador=',', encoding='utf-8')
base_extrema_pobreza.head()

Importando o dataframe: ../../avaliacao-python-ciencia-dados/Bases/gm_extreme_poverty_rate.csv ...
Dataframe importado com sucesso! 
Dimensão: (190, 303)


geo         name   1800   1801   1802   1803   1804   1805   1806   1807  \
0  afg  Afghanistan  91.28  91.28  91.28  91.28  91.28  91.28  91.28  91.28   
1  ago       Angola  91.81  91.81  91.81  91.81  91.81  91.33  91.33  91.33   
2  alb      Albania  92.42  92.42  92.42  92.40  92.40  92.01  92.01  92.01   
3  are          UAE  30.60  29.50  29.50  29.50  29.59  29.59  29.37  29.37   
4  arg    Argentina  38.64  37.00  37.01  36.49  36.27  36.27  36.04  36.04   

   ...  2091  2092  2093  2094  2095  2096  2097  2098  2099  2100  
0  ...  4.57  4.49  4.15  3.83  3.55  3.28  2.97  2.71  2.68  2.39  
1  ...  1.78  1.64  1.51  1.50  1.39  1.29  1.28  1.18  1.09  1.09  
2  ...  0.11  0.10  0.10  0.10  0.10  0.10  0.10  0.10  0.10  0.09  
3  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  ...  0.10  0.10  0.10  0.09  0.09  0.09  0.09  0.09  0.09  0.08  

[5 rows x 303 columns]

In [21]:
#Transposição da base
#Redução de casas decimais
#Ajuste do nome das colunas
base_extrema_pobreza = transpoe_e_trata_dataframe(base_extrema_pobreza, nome_coluna_resultado='extremePovertyRate')
base_extrema_pobreza.head()

Removendo dados duplicados ...
Foram encontrados 0 registros duplicados

Prévia dos dados duplicados:
Dados duplicados removidos com sucesso! 
Dimensão: (53916, 3)


country  year  extremePovertyRate
0  Afghanistan  1800               91.28
1  Afghanistan  1801               91.28
2  Afghanistan  1802               91.28
3  Afghanistan  1803               91.28
4  Afghanistan  1804               91.28

In [22]:
juncao = junta_dataframes(juncao, base_extrema_pobreza, chave=['country', 'year'], tipo_juncao='left')
juncao.head()

Dataframes unidos com sucesso! 
Dimensão: (1704, 14)


country  year         pop continent  lifeExp  gdpPercap  \
0  Afghanistan  1952   8.425.333      Asia     28.8     779.45   
1  Afghanistan  1957   9.240.934      Asia     30.3     820.85   
2  Afghanistan  1962  10.267.083      Asia     32.0     853.10   
3  Afghanistan  1967  11.537.966      Asia     34.0     836.20   
4  Afghanistan  1972  13.079.460      Asia     36.1     739.98   

   oneSidedViolence  nonStateConflict  extrasystemic  intrastate  interstate  \
0               NaN               NaN            NaN         NaN         NaN   
1               NaN               NaN            NaN         NaN         NaN   
2               NaN               NaN            NaN         NaN         NaN   
3               NaN               NaN            NaN         NaN         NaN   
4               NaN               NaN            NaN         NaN         NaN   

   totalWarDeaths  gdpPerCaptaInflactionAdjusted  extremePovertyRate  
0             NaN                        1540.11               43.70  
1             NaN                        1708.81               37.13  
2             NaN                        1869.50               30.54  
3             NaN                        1930.54               33.47  
4             NaN                        1602.84               43.66

In [23]:
#Adicionando a base de mortalidade infantil
base_mortalidade_infantil = importa_csv(dataframe_name='../../avaliacao-python-ciencia-dados/Bases/child_mortality_0_5_year_olds_dying_per_1000_born.csv', separador=',', encoding='utf-8')
base_mortalidade_infantil.head()

Importando o dataframe: ../../avaliacao-python-ciencia-dados/Bases/child_mortality_0_5_year_olds_dying_per_1000_born.csv ...
Dataframe importado com sucesso! 
Dimensão: (192, 303)


geo         name    1800    1801    1802    1803    1804    1805    1806  \
0  afg  Afghanistan  468.58  468.58  468.58  468.58  468.58  468.58  469.98   
1  ago       Angola  485.68  485.68  485.68  485.68  485.68  485.68  485.68   
2  alb      Albania     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3  and      Andorra     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4  are          UAE     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

     1807  ...     2091     2092     2093     2094     2095     2096     2097  \
0  469.98  ...  10.6577  10.4730  10.2963  10.1103   9.9446   9.7574   9.5760   
1  485.68  ...  23.1019  22.8020  22.5032  22.1450  21.8240  21.5334  21.2922   
2     NaN  ...   1.8035   1.7809   1.7565   1.7306   1.7090   1.6871   1.6672   
3     NaN  ...   1.0973   1.0704   1.0470   1.0227   0.9977   0.9737   0.9492   
4     NaN  ...   1.1265   1.1064   1.0879   1.0685   1.0476   1.0305   1.0133   

      2098     2099     2100  
0   9.4139   9.2223   9.0653  
1  21.0403  20.8104  20.5798  
2   1.6437   1.6208   1.6005  
3   0.9264   0.9049   0.8830  
4   0.9970   0.9795   0.9624  

[5 rows x 303 columns]

In [24]:
base_mortalidade_infantil = transpoe_e_trata_dataframe(base_mortalidade_infantil, nome_coluna_resultado='childMortalityPer1000Born')
base_mortalidade_infantil.head()

Removendo dados duplicados ...
Foram encontrados 0 registros duplicados

Prévia dos dados duplicados:
Dados duplicados removidos com sucesso! 
Dimensão: (51880, 3)


country  year  childMortalityPer1000Born
0  Afghanistan  1800                     468.58
1  Afghanistan  1801                     468.58
2  Afghanistan  1802                     468.58
3  Afghanistan  1803                     468.58
4  Afghanistan  1804                     468.58

In [25]:
juncao = junta_dataframes(juncao, base_mortalidade_infantil, chave=['country', 'year'], tipo_juncao='left')
juncao.head()

Dataframes unidos com sucesso! 
Dimensão: (1704, 15)


country  year         pop continent  lifeExp  gdpPercap  \
0  Afghanistan  1952   8.425.333      Asia     28.8     779.45   
1  Afghanistan  1957   9.240.934      Asia     30.3     820.85   
2  Afghanistan  1962  10.267.083      Asia     32.0     853.10   
3  Afghanistan  1967  11.537.966      Asia     34.0     836.20   
4  Afghanistan  1972  13.079.460      Asia     36.1     739.98   

   oneSidedViolence  nonStateConflict  extrasystemic  intrastate  interstate  \
0               NaN               NaN            NaN         NaN         NaN   
1               NaN               NaN            NaN         NaN         NaN   
2               NaN               NaN            NaN         NaN         NaN   
3               NaN               NaN            NaN         NaN         NaN   
4               NaN               NaN            NaN         NaN         NaN   

   totalWarDeaths  gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
0             NaN                        1540.11               43.70   
1             NaN                        1708.81               37.13   
2             NaN                        1869.50               30.54   
3             NaN                        1930.54               33.47   
4             NaN                        1602.84               43.66   

   childMortalityPer1000Born  
0                     408.55  
1                     375.91  
2                     345.84  
3                     318.37  
4                     290.73

In [26]:
#Adicionando Arquivo de densidade populacional
base_densidade_populacional = importa_csv(dataframe_name='../../avaliacao-python-ciencia-dados/Bases/population_density_per_square_km.csv', separador=',', encoding='utf-8')
base_densidade_populacional.head()

Importando o dataframe: ../../avaliacao-python-ciencia-dados/Bases/population_density_per_square_km.csv ...
Dataframe importado com sucesso! 
Dimensão: (195, 153)


geo         name    1950    1951    1952    1953    1954    1955    1956  \
0  afg  Afghanistan  11.976  12.135  12.302  12.470  12.641  12.825  13.021   
1  ago       Angola   3.651   3.692   3.731   3.775   3.823   3.876   3.934   
2  alb      Albania  45.545  46.586  47.754  49.039  50.408  51.787  53.172   
3  and      Andorra  12.757  12.380  11.607  11.326  11.919  13.118  14.393   
4  are          UAE   1.049   1.120   1.193   1.269   1.347   1.426   1.506   

     1957  ...     2091     2092     2093     2094     2095     2096     2097  \
0  13.227  ...  191.145  192.292  193.424  194.529  195.581  196.594  197.607   
1   3.995  ...  111.430  112.496  113.540  114.571  115.587  116.580  117.550   
2  54.674  ...   49.057   48.332   47.627   46.943   46.281   45.639   45.016   
3  15.744  ...  112.299  110.923  109.540  108.177  106.832  105.501  104.198   
4   1.588  ...  338.973  342.111  345.263  348.432  351.621  354.832  358.069   

      2098     2099     2100  
0  198.622  199.626  200.551  
1  118.508  119.446  120.354  
2   44.411   43.824   43.251  
3  102.920  101.682  100.473  
4  361.336  364.633  367.964  

[5 rows x 153 columns]

In [27]:
base_densidade_populacional = transpoe_e_trata_dataframe(base_densidade_populacional, nome_coluna_resultado='popDensityPerSquareKm')
base_densidade_populacional.head()

Removendo dados duplicados ...
Foram encontrados 0 registros duplicados

Prévia dos dados duplicados:
Dados duplicados removidos com sucesso! 
Dimensão: (29445, 3)


country  year  popDensityPerSquareKm
0  Afghanistan  1950                  11.98
1  Afghanistan  1951                  12.14
2  Afghanistan  1952                  12.30
3  Afghanistan  1953                  12.47
4  Afghanistan  1954                  12.64

In [28]:
juncao = junta_dataframes(juncao, base_densidade_populacional, chave=['country', 'year'], tipo_juncao='left')
juncao.head()

Dataframes unidos com sucesso! 
Dimensão: (1704, 16)


country  year         pop continent  lifeExp  gdpPercap  \
0  Afghanistan  1952   8.425.333      Asia     28.8     779.45   
1  Afghanistan  1957   9.240.934      Asia     30.3     820.85   
2  Afghanistan  1962  10.267.083      Asia     32.0     853.10   
3  Afghanistan  1967  11.537.966      Asia     34.0     836.20   
4  Afghanistan  1972  13.079.460      Asia     36.1     739.98   

   oneSidedViolence  nonStateConflict  extrasystemic  intrastate  interstate  \
0               NaN               NaN            NaN         NaN         NaN   
1               NaN               NaN            NaN         NaN         NaN   
2               NaN               NaN            NaN         NaN         NaN   
3               NaN               NaN            NaN         NaN         NaN   
4               NaN               NaN            NaN         NaN         NaN   

   totalWarDeaths  gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
0             NaN                        1540.11               43.70   
1             NaN                        1708.81               37.13   
2             NaN                        1869.50               30.54   
3             NaN                        1930.54               33.47   
4             NaN                        1602.84               43.66   

   childMortalityPer1000Born  popDensityPerSquareKm  
0                     408.55                  12.30  
1                     375.91                  13.23  
2                     345.84                  14.48  
3                     318.37                  16.18  
4                     290.73                  18.26

In [29]:
#Adicionando a base de desenvolvimento humano (HDI)
base_hdi = importa_csv(dataframe_name='../../avaliacao-python-ciencia-dados/Bases/hdi_human_development_index.csv', separador=',', encoding='utf-8')
base_hdi.head()

Importando o dataframe: ../../avaliacao-python-ciencia-dados/Bases/hdi_human_development_index.csv ...
Dataframe importado com sucesso! 
Dimensão: (192, 36)


geo         name   1990   1991   1992   1993   1994   1995   1996   1997  \
0  afg  Afghanistan  0.285  0.291  0.301  0.311  0.305  0.329  0.334  0.338   
1  ago       Angola    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2  alb      Albania  0.654  0.638  0.622  0.624  0.629  0.638  0.647  0.645   
3  and      Andorra    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4  are          UAE  0.713  0.724  0.729  0.736  0.742  0.747  0.753  0.761   

   ...   2014   2015   2016   2017   2018   2019   2020   2021   2022   2023  
0  ...  0.497  0.496  0.495  0.496  0.498  0.507  0.501  0.486  0.495  0.496  
1  ...  0.577  0.603  0.609  0.610  0.611  0.611  0.610  0.609  0.615  0.616  
2  ...  0.797  0.797  0.797  0.798  0.801  0.805  0.794  0.794  0.806  0.810  
3  ...  0.866  0.869  0.872  0.873  0.875  0.876  0.851  0.871  0.893  0.913  
4  ...  0.853  0.857  0.861  0.884  0.901  0.915  0.909  0.903  0.921  0.940  

[5 rows x 36 columns]

In [30]:
base_hdi = transpoe_e_trata_dataframe(base_hdi, nome_coluna_resultado='humanDevelopmentIndex')
#base_hdi['humanDevelopmentIndex'] = base_hdi['humanDevelopmentIndex'].mul(100).map("{:.0f}%".format)
# Eu ia converter, mas pesquisando, o HDI é normalmente representado como um número decimal entre 0 e 1 mas não necessariamente é um indicador de porcentagem, só é do calculo mesmo, então vou deixar assim.
base_hdi.head()

Removendo dados duplicados ...
Foram encontrados 0 registros duplicados

Prévia dos dados duplicados:
Dados duplicados removidos com sucesso! 
Dimensão: (5906, 3)


country  year  humanDevelopmentIndex
0  Afghanistan  1990                   0.28
1  Afghanistan  1991                   0.29
2  Afghanistan  1992                   0.30
3  Afghanistan  1993                   0.31
4  Afghanistan  1994                   0.30

In [31]:
juncao = junta_dataframes(juncao, base_hdi, chave=['country', 'year'], tipo_juncao='left')
juncao.head()

Dataframes unidos com sucesso! 
Dimensão: (1704, 17)


country  year         pop continent  lifeExp  gdpPercap  \
0  Afghanistan  1952   8.425.333      Asia     28.8     779.45   
1  Afghanistan  1957   9.240.934      Asia     30.3     820.85   
2  Afghanistan  1962  10.267.083      Asia     32.0     853.10   
3  Afghanistan  1967  11.537.966      Asia     34.0     836.20   
4  Afghanistan  1972  13.079.460      Asia     36.1     739.98   

   oneSidedViolence  nonStateConflict  extrasystemic  intrastate  interstate  \
0               NaN               NaN            NaN         NaN         NaN   
1               NaN               NaN            NaN         NaN         NaN   
2               NaN               NaN            NaN         NaN         NaN   
3               NaN               NaN            NaN         NaN         NaN   
4               NaN               NaN            NaN         NaN         NaN   

   totalWarDeaths  gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
0             NaN                        1540.11               43.70   
1             NaN                        1708.81               37.13   
2             NaN                        1869.50               30.54   
3             NaN                        1930.54               33.47   
4             NaN                        1602.84               43.66   

   childMortalityPer1000Born  popDensityPerSquareKm  humanDevelopmentIndex  
0                     408.55                  12.30                    NaN  
1                     375.91                  13.23                    NaN  
2                     345.84                  14.48                    NaN  
3                     318.37                  16.18                    NaN  
4                     290.73                  18.26                    NaN

Agora iniciaremos uma análise exploratória básica do projeto, com o objetivo de ver se tem mais algo que se destaque:

In [32]:
juncao.describe()

year      lifeExp      gdpPercap  oneSidedViolence  \
count  1704.00000  1704.000000    1704.000000        520.000000   
mean   1979.50000    59.475235    7215.327095       1843.396154   
std      17.26533    12.916468    9857.454534      34328.028878   
min    1952.00000    23.600000     241.170000          0.000000   
25%    1965.75000    48.200000    1202.060000          0.000000   
50%    1979.50000    60.750000    3531.845000          0.000000   
75%    1993.25000    70.800000    9325.462500         46.500000   
max    2007.00000    82.600000  113523.130000     782261.000000   

       nonStateConflict  extrasystemic    intrastate    interstate  \
count        520.000000          520.0    520.000000    520.000000   
mean         152.476923            0.0   1065.101923    256.819231   
std          569.971029            0.0   4053.362937   4322.069164   
min            0.000000            0.0      0.000000      0.000000   
25%            0.000000            0.0      0.000000      0.000000   
50%            0.000000            0.0      0.000000      0.000000   
75%            0.000000            0.0    100.500000      0.000000   
max         5817.000000            0.0  43871.000000  95934.000000   

       totalWarDeaths  gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
count      520.000000                    1548.000000         1482.000000   
mean      3317.794231                    9536.118630           29.564143   
std      35009.273292                   11606.065523           29.217754   
min          0.000000                     433.690000            0.010000   
25%          0.000000                    1918.902500            1.507500   
50%          0.000000                    4657.985000           18.875000   
75%        372.750000                   12143.375000           54.152500   
max     785697.000000                   88925.070000           98.380000   

       childMortalityPer1000Born  popDensityPerSquareKm  humanDevelopmentIndex  
count                1521.000000            1548.000000             468.000000  
mean                  116.850302             112.672720               0.633141  
std                   100.116115             379.720921               0.177727  
min                     2.850000               0.520000               0.220000  
25%                    28.350000              14.285000               0.480000  
50%                    87.970000              42.765000               0.660000  
75%                   185.910000             104.460000               0.772500  
max                   451.010000            6729.890000               0.940000

Alguns pontos de destaque na análise:
1. A coluna extrasystemic está vazia, pode ser removida.
2. Apesar de termos um ano extremamente baixo ali no mínimo (23 anos) a média ainda se manteve em 59 anos, com a mediana ali em 60, o que indica que o valor mínimo é um outlier (justificado pela guerra em Rwanda)
3. no caso da renda per capta, temos uma flutuação absurda entre o valor mínimo e máximo, o que indica um valor que vale a pena analisar mais a fundo
4. Em todos os casos de conflitos mapeados, a mediana foi zero, o que indica que, pelo menos, metade da base é "zerada".
5. gdpPerCaptaInflactionAdjusted -> tem 1548 registros, aproximadamente 200 a menos que a base original, vale analisar o pq.
6. extremePovertyRate -> tem 1482 registros, vale analisar o pq. O valor encontra-se entre 0 e 100, o que faz sentido dado que é um resultado percentual
7. childMortalityPer1000Born ->  tem 1521 registros, vale analisar o pq.
8. popDensityPerSquareKm ->  tem 1548 registros, vale analisar o pq.
9. hdi ->  tem 468 registros, vale analisar o pq. O valor está entre 0 e 1, o que é esperado dado o indicador. 

Agora vamos analisar os indicadores citados acima que parecem esquisitos e remover a coluna vazia.

In [33]:
# Analisar o caso da Rwanda pra ver o outlier mais de perto:
filtro = juncao['country'] == 'Rwanda'
ausentes = juncao[filtro].index
juncao.loc[ausentes].head(20)

country  year        pop continent  lifeExp  gdpPercap  oneSidedViolence  \
1284  Rwanda  1952  2.534.927    Africa     40.0     493.32               NaN   
1285  Rwanda  1957  2.822.082    Africa     41.5     540.29               NaN   
1286  Rwanda  1962  3.051.242    Africa     43.0     597.47               NaN   
1287  Rwanda  1967  3.451.079    Africa     44.1     510.96               NaN   
1288  Rwanda  1972  3.992.121    Africa     44.6     590.58               NaN   
1289  Rwanda  1977  4.657.072    Africa     45.0     670.08               NaN   
1290  Rwanda  1982  5.507.565    Africa     46.2     881.57               NaN   
1291  Rwanda  1987  6.349.365    Africa     44.0     847.99               NaN   
1292  Rwanda  1992  7.290.203    Africa     23.6     737.07            3013.0   
1293  Rwanda  1997  7.212.583    Africa     36.1     589.94          782261.0   
1294  Rwanda  2002  7.852.401    Africa     43.4     785.65             708.0   
1295  Rwanda  2007  8.860.588    Africa     46.2     863.09               0.0   

      nonStateConflict  extrasystemic  intrastate  interstate  totalWarDeaths  \
1284               NaN            NaN         NaN         NaN             NaN   
1285               NaN            NaN         NaN         NaN             NaN   
1286               NaN            NaN         NaN         NaN             NaN   
1287               NaN            NaN         NaN         NaN             NaN   
1288               NaN            NaN         NaN         NaN             NaN   
1289               NaN            NaN         NaN         NaN             NaN   
1290               NaN            NaN         NaN         NaN             NaN   
1291               NaN            NaN         NaN         NaN             NaN   
1292               0.0            0.0      1366.0         0.0          4379.0   
1293              92.0            0.0      3344.0         0.0        785697.0   
1294               0.0            0.0      3139.0         0.0          3847.0   
1295               0.0            0.0         0.0         0.0             0.0   

      gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
1284                        1135.91               75.50   
1285                        1205.18               72.67   
1286                        1271.78               70.77   
1287                        1036.86               78.70   
1288                        1110.96               76.92   
1289                        1168.87               75.07   
1290                        1270.34               70.34   
1291                        1278.75               72.24   
1292                        1042.52               74.54   
1293                         783.39               78.74   
1294                        1042.08               69.24   
1295                        1263.54               64.10   

      childMortalityPer1000Born  popDensityPerSquareKm  humanDevelopmentIndex  
1284                     257.93                  95.68                    NaN  
1285                     242.09                 113.50                    NaN  
1286                     212.41                 131.60                    NaN  
1287                     211.85                 149.92                    NaN  
1288                     227.99                 174.32                    NaN  
1289                     250.95                 201.91                    NaN  
1290                     186.14                 235.92                    NaN  
1291                     150.30                 278.79                    NaN  
1292                     173.21                 317.36                    NaN  
1293                     220.75                 320.33                   0.30  
1294                     149.60                 346.97                   0.38  
1295                      84.47                 394.46                   0.47

Podemos ver ali, nos anos de 1992 e 1997, houve algum conflito que levou a uma morte massiva no país. Procurando na internet, chegamos à psotagem abaixo:
> _"Durante os 100 dias seguintes, entre abril e julho de 1994, ocorreu um massacre sistemático, com cerca de 800 mil a 1 milhão de pessoas mortas, principalmente tutsis, por meio de facões e outras formas de violência."_

Isso justifica o outlier na tabela mesmo olhando pra expectativa de vida de Rwanda, dado que a gueraa civil começou em 1990 e atigiu o ápice em 1994, com essa volumetria exuberante de mortos.

Agora olharemos para a renda per capta, item 3:

In [34]:
juncao.sort_values(by='gdpPercap', ascending=True).head(50)

country  year          pop continent  lifeExp  gdpPercap  \
334     Congo Dem. Rep.  2002   55.379.852    Africa     45.0     241.17   
335     Congo Dem. Rep.  2007   64.606.759    Africa     46.5     277.55   
876             Lesotho  1952      748.747    Africa     42.1     298.85   
624       Guinea-Bissau  1952      580.653    Africa     32.5     299.85   
333     Congo Dem. Rep.  1997   47.798.986    Africa     42.6     312.19   
492             Eritrea  1952    1.438.760    Africa     35.9     328.94   
1044            Myanmar  1952   20.092.996      Asia     36.3     331.00   
877             Lesotho  1957      813.338    Africa     45.0     336.00   
204             Burundi  1952    2.445.618    Africa     39.0     339.30   
493             Eritrea  1957    1.542.611    Africa     38.0     344.16   
1052            Myanmar  1992   40.546.538      Asia     59.3     347.00   
1047            Myanmar  1967   25.870.271      Asia     49.4     349.00   
1045            Myanmar  1957   21.731.844      Asia     41.9     350.00   
206             Burundi  1962    2.961.915    Africa     42.0     355.20   
1048            Myanmar  1972   28.466.390      Asia     53.1     357.00   
504            Ethiopia  1952   20.860.941    Africa     34.1     362.15   
216            Cambodia  1952    4.693.836      Asia     39.4     368.47   
924              Malawi  1952    2.917.802    Africa     36.3     369.17   
1049            Myanmar  1977   31.528.087      Asia     56.1     371.00   
480   Equatorial Guinea  1952      216.964    Africa     34.5     375.64   
505            Ethiopia  1957   22.815.614    Africa     36.7     378.90   
205             Burundi  1957    2.667.518    Africa     40.5     379.56   
494             Eritrea  1962    1.666.618    Africa     40.2     381.00   
1051            Myanmar  1987   38.028.578      Asia     58.3     385.00   
1046            Myanmar  1962   23.634.436      Asia     45.1     388.00   
1039         Mozambique  1987   12.891.952    Africa     42.9     389.88   
288               China  1952  556.263.527      Asia     44.0     400.45   
1692           Zimbabwe  1952    3.080.907    Africa     48.5     406.88   
1040         Mozambique  1992   13.160.731    Africa     44.3     410.90   
878             Lesotho  1962      893.143    Africa     47.7     411.80   
207             Burundi  1967    3.330.989    Africa     43.5     412.98   
899             Liberia  2007    3.193.942    Africa     45.7     414.51   
1053            Myanmar  1997   43.247.867      Asia     60.3     415.00   
925              Malawi  1957    3.221.238    Africa     37.2     416.37   
506            Ethiopia  1962   25.145.372    Africa     40.1     419.46   
512            Ethiopia  1992   52.088.559    Africa     48.1     421.35   
220            Cambodia  1972    7.450.606      Asia     40.3     421.62   
1050            Myanmar  1982   34.680.442      Asia     58.1     424.00   
481   Equatorial Guinea  1957      232.922    Africa     36.0     426.10   
926              Malawi  1962    3.628.608    Africa     38.4     427.90   
215             Burundi  2007    8.390.505    Africa     49.6     430.07   
625       Guinea-Bissau  1957      601.095    Africa     33.5     431.79   
217            Cambodia  1957    5.322.536      Asia     41.4     434.04   
214             Burundi  2002    7.021.078    Africa     47.4     446.40   
948                Mali  1952    3.838.168    Africa     33.7     452.34   
332     Congo Dem. Rep.  1992   41.672.143    Africa     45.5     457.72   
1038         Mozambique  1982   12.587.223    Africa     42.8     462.21   
213             Burundi  1997    6.121.610    Africa     45.3     463.12   
208             Burundi  1972    3.529.983    Africa     44.1     464.10   
1032         Mozambique  1952    6.446.316    Africa     31.3     468.53   

      oneSidedViolence  nonStateConflict  extrasystemic  intrastate  \
334                NaN               NaN            NaN         NaN 

Como o valor da média ali tá sendo visto sem levar em conta o ano, apenas o indicador como um todo, a análise fica meio limitada, valendo aprofundar mais no próximo item um comparativo por ano específico. Nada de esquisito nos dados aqui.

Agora vamos ao item 5:

In [35]:
gdpPerCaptanulo = juncao[juncao['gdpPerCaptaInflactionAdjusted'].isnull()]
gdpPerCaptanulo.info()
gdpPerCaptanulo['country'].drop_duplicates().head(100)

<class 'pandas.core.frame.DataFrame'>
Index: 156 entries, 324 to 1679
Data columns (total 17 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   country                        156 non-null    object 
 1   year                           156 non-null    int64  
 2   pop                            156 non-null    object 
 3   continent                      156 non-null    object 
 4   lifeExp                        156 non-null    float64
 5   gdpPercap                      156 non-null    float64
 6   oneSidedViolence               12 non-null     float64
 7   nonStateConflict               12 non-null     float64
 8   extrasystemic                  12 non-null     float64
 9   intrastate                     12 non-null     float64
 10  interstate                     12 non-null     float64
 11  totalWarDeaths                 12 non-null     float64
 12  gdpPerCaptaInflactionAdjusted  0 non-null      float

324        Congo Dem. Rep.
336             Congo Rep.
660        Hong Kong China
828        Korea Dem. Rep.
840             Korea Rep.
1248           Puerto Rico
1260               Reunion
1452             Swaziland
1500                Taiwan
1596        United Kingdom
1608         United States
1656    West Bank and Gaza
1668            Yemen Rep.
Name: country, dtype: object

Temos 13 paises esquisitos:
- Congo Dem. Rep.
- Congo Rep.
- Hong Kong China
- Korea Dem. Rep.
- Korea Rep.
- Puerto Rico
- Reunion
- Swaziland
- Taiwan
- United Kingdom
- United States
- West Bank and Gaza
- Yemen Rep.

Analisando, a justificativa é: 
- Congo Dem. Rep. -> nome na tabela é: Congo, Dem. Rep.
- Congo Rep. -> nome na tabela é: Congo, Rep.
- Hong Kong China -> Não existe na tabela
- Korea Dem. Rep. -> South Korea
- Korea Rep. -> nome na tabela é: North Korea
- Puerto Rico -> Não existe na tabela
- Reunion ->  Não existe na tabela, provavelmente tratado na França
- Swaziland ->  nome na tabela é: Eswatini
- Taiwan -> Não existe na tabela, provavelmente tratado na China
- United Kingdom ->  nome na tabela é: UK
- United States ->  nome na tabela é: USA
- West Bank and Gaza ->  nome na tabela é: Palestine
- Yemen Rep. ->  nome na tabela é: Yemen

Com isso, vamos dropar essa coluna, corrigir a base original e refazer o join: (Sei que não é o melhor jeito, dava pra corrigir lá em cima antes de fazer o join, mas ai a gente ia perder o processo, que entendo que é o que você quer capturar aqui, professor, então fiz do jeito mais "burro" ou mais custoso aqui no caso)

In [36]:

renomeia_coluna(juncao, nome_anterior='gdpPerCaptaInflactionAdjusted', novo_nome='gdpPerCapitaInflationAdjusted_old', inplace=True)
juncao.head()

country  year         pop continent  lifeExp  gdpPercap  \
0  Afghanistan  1952   8.425.333      Asia     28.8     779.45   
1  Afghanistan  1957   9.240.934      Asia     30.3     820.85   
2  Afghanistan  1962  10.267.083      Asia     32.0     853.10   
3  Afghanistan  1967  11.537.966      Asia     34.0     836.20   
4  Afghanistan  1972  13.079.460      Asia     36.1     739.98   

   oneSidedViolence  nonStateConflict  extrasystemic  intrastate  interstate  \
0               NaN               NaN            NaN         NaN         NaN   
1               NaN               NaN            NaN         NaN         NaN   
2               NaN               NaN            NaN         NaN         NaN   
3               NaN               NaN            NaN         NaN         NaN   
4               NaN               NaN            NaN         NaN         NaN   

   totalWarDeaths  gdpPerCapitaInflationAdjusted_old  extremePovertyRate  \
0             NaN                            1540.11               43.70   
1             NaN                            1708.81               37.13   
2             NaN                            1869.50               30.54   
3             NaN                            1930.54               33.47   
4             NaN                            1602.84               43.66   

   childMortalityPer1000Born  popDensityPerSquareKm  humanDevelopmentIndex  
0                     408.55                  12.30                    NaN  
1                     375.91                  13.23                    NaN  
2                     345.84                  14.48                    NaN  
3                     318.37                  16.18                    NaN  
4                     290.73                  18.26                    NaN

In [37]:

base_gdp_ajustado = corrige_paises_estranhos(base_gdp_ajustado)

juncao = junta_dataframes(juncao, base_gdp_ajustado, chave=['country', 'year'], tipo_juncao='left')
juncao.describe()

Dataframes unidos com sucesso! 
Dimensão: (1704, 18)


year      lifeExp      gdpPercap  oneSidedViolence  \
count  1704.00000  1704.000000    1704.000000        520.000000   
mean   1979.50000    59.475235    7215.327095       1843.396154   
std      17.26533    12.916468    9857.454534      34328.028878   
min    1952.00000    23.600000     241.170000          0.000000   
25%    1965.75000    48.200000    1202.060000          0.000000   
50%    1979.50000    60.750000    3531.845000          0.000000   
75%    1993.25000    70.800000    9325.462500         46.500000   
max    2007.00000    82.600000  113523.130000     782261.000000   

       nonStateConflict  extrasystemic    intrastate    interstate  \
count        520.000000          520.0    520.000000    520.000000   
mean         152.476923            0.0   1065.101923    256.819231   
std          569.971029            0.0   4053.362937   4322.069164   
min            0.000000            0.0      0.000000      0.000000   
25%            0.000000            0.0      0.000000      0.000000   
50%            0.000000            0.0      0.000000      0.000000   
75%            0.000000            0.0    100.500000      0.000000   
max         5817.000000            0.0  43871.000000  95934.000000   

       totalWarDeaths  gdpPerCapitaInflationAdjusted_old  extremePovertyRate  \
count      520.000000                        1548.000000         1482.000000   
mean      3317.794231                        9536.118630           29.564143   
std      35009.273292                       11606.065523           29.217754   
min          0.000000                         433.690000            0.010000   
25%          0.000000                        1918.902500            1.507500   
50%          0.000000                        4657.985000           18.875000   
75%        372.750000                       12143.375000           54.152500   
max     785697.000000                       88925.070000           98.380000   

       childMortalityPer1000Born  popDensityPerSquareKm  \
count                1521.000000            1548.000000   
mean                  116.850302             112.672720   
std                   100.116115             379.720921   
min                     2.850000               0.520000   
25%                    28.350000              14.285000   
50%                    87.970000              42.765000   
75%                   185.910000             104.460000   
max                   451.010000            6729.890000   

       humanDevelopmentIndex  gdpPerCaptaInflactionAdjusted  
count             468.000000                    1656.000000  
mean                0.633141                    9534.027319  
std                 0.177727                   11681.196053  
min                 0.220000                     433.690000  
25%                 0.480000                    1913.582500  
50%                 0.660000                    4581.085000  
75%                 0.772500                   12199.975000  
max                 0.940000                   88925.070000

In [38]:
juncao = juncao.drop(columns=['gdpPerCapitaInflationAdjusted_old'])

Pronto, colunas renomeadas, vamos repetir o processo pro item 6:

In [39]:
pobrezanulo = juncao[juncao['extremePovertyRate'].isnull()]
pobrezanulo.head(100)

country  year        pop continent  lifeExp  gdpPercap  \
185  Bulgaria  1977  8.797.022    Europe     70.8    7612.24   
186  Bulgaria  1982  8.892.098    Europe     71.1    8224.19   
187  Bulgaria  1987  8.971.958    Europe     71.3    8239.85   
188  Bulgaria  1992  8.658.506    Europe     71.2    6302.62   
189  Bulgaria  1997  8.066.057    Europe     70.3    5970.39   
..        ...   ...        ...       ...      ...        ...   
857    Kuwait  1977  1.140.357      Asia     69.3   59265.48   
861    Kuwait  1997  1.765.345      Asia     76.2   40300.62   
862    Kuwait  2002  2.111.561      Asia     76.9   35110.11   
863    Kuwait  2007  2.505.559      Asia     77.6   47306.99   
873   Lebanon  1997  3.430.388      Asia     70.3    8754.96   

     oneSidedViolence  nonStateConflict  extrasystemic  intrastate  \
185               NaN               NaN            NaN         NaN   
186               NaN               NaN            NaN         NaN   
187               NaN               NaN            NaN         NaN   
188               0.0               0.0            0.0         0.0   
189               0.0               0.0            0.0         0.0   
..                ...               ...            ...         ...   
857               NaN               NaN            NaN         NaN   
861               1.0               0.0            0.0         0.0   
862               0.0               0.0            0.0         0.0   
863               0.0               0.0            0.0         0.0   
873             103.0             110.0            0.0       456.0   

     interstate  totalWarDeaths  extremePovertyRate  \
185         NaN             NaN                 NaN   
186         NaN             NaN                 NaN   
187         NaN             NaN                 NaN   
188         0.0             0.0                 NaN   
189         0.0             0.0                 NaN   
..          ...             ...                 ...   
857         NaN             NaN                 NaN   
861         0.0             1.0                 NaN   
862         0.0             0.0                 NaN   
863         0.0             0.0                 NaN   
873         0.0           669.0                 NaN   

     childMortalityPer1000Born  popDensityPerSquareKm  humanDevelopmentIndex  \
185                      27.25                  81.23                    NaN   
186                      22.20                  82.22                    NaN   
187                      17.95                  82.81                    NaN   
188                      19.24                  78.96                   0.71   
189                      19.41                  75.81                   0.71   
..                         ...                    ...                    ...   
857                      44.81                  70.60                    NaN   
861                      13.80                 100.33                   0.76   
862                      12.35                 115.61                   0.78   
863                      11.43                 140.75                   0.80   
873                      23.33                 401.60                    NaN   

     gdpPerCaptaInflactionAdjusted  
185                        9553.12  
186                       11282.34  
187                       12751.02  
188                       10330.88  
189                       10129.72  
..                             ...  
857                       51804.87  
861                       51842.44  
862                       52384.22  
863                       71487.51  
873                       11584.82  

[100 rows x 17 columns]

Aqui existe um detalhe a mais: Temos países que não temos dados pra anos em específico, então não trataremos esses casos, trataremos apenas os países já conhecidos.

In [40]:
renomeia_coluna(juncao, nome_anterior='extremePovertyRate', novo_nome='extremePovertyRate_old', inplace=True)

base_extrema_pobreza = corrige_paises_estranhos(base_extrema_pobreza)

juncao = junta_dataframes(juncao, base_extrema_pobreza, chave=['country', 'year'], tipo_juncao='left')
juncao = juncao.drop(columns=['extremePovertyRate_old'])
juncao.describe()

Dataframes unidos com sucesso! 
Dimensão: (1704, 18)


year      lifeExp      gdpPercap  oneSidedViolence  \
count  1704.00000  1704.000000    1704.000000        520.000000   
mean   1979.50000    59.475235    7215.327095       1843.396154   
std      17.26533    12.916468    9857.454534      34328.028878   
min    1952.00000    23.600000     241.170000          0.000000   
25%    1965.75000    48.200000    1202.060000          0.000000   
50%    1979.50000    60.750000    3531.845000          0.000000   
75%    1993.25000    70.800000    9325.462500         46.500000   
max    2007.00000    82.600000  113523.130000     782261.000000   

       nonStateConflict  extrasystemic    intrastate    interstate  \
count        520.000000          520.0    520.000000    520.000000   
mean         152.476923            0.0   1065.101923    256.819231   
std          569.971029            0.0   4053.362937   4322.069164   
min            0.000000            0.0      0.000000      0.000000   
25%            0.000000            0.0      0.000000      0.000000   
50%            0.000000            0.0      0.000000      0.000000   
75%            0.000000            0.0    100.500000      0.000000   
max         5817.000000            0.0  43871.000000  95934.000000   

       totalWarDeaths  childMortalityPer1000Born  popDensityPerSquareKm  \
count      520.000000                1521.000000            1548.000000   
mean      3317.794231                 116.850302             112.672720   
std      35009.273292                 100.116115             379.720921   
min          0.000000                   2.850000               0.520000   
25%          0.000000                  28.350000              14.285000   
50%          0.000000                  87.970000              42.765000   
75%        372.750000                 185.910000             104.460000   
max     785697.000000                 451.010000            6729.890000   

       humanDevelopmentIndex  gdpPerCaptaInflactionAdjusted  \
count             468.000000                    1656.000000   
mean                0.633141                    9534.027319   
std                 0.177727                   11681.196053   
min                 0.220000                     433.690000   
25%                 0.480000                    1913.582500   
50%                 0.660000                    4581.085000   
75%                 0.772500                   12199.975000   
max                 0.940000                   88925.070000   

       extremePovertyRate  
count         1589.000000  
mean            29.899880  
std             29.521719  
min              0.010000  
25%              1.440000  
50%             19.060000  
75%             54.910000  
max             99.300000

Agora vamos ao item 7:

In [ ]:
renomeia_coluna(juncao, nome_anterior='childMortalityPer1000Born', novo_nome='childMortalityPer1000Born_old', inplace=True)

base_mortalidade_infantil = corrige_paises_estranhos(base_mortalidade_infantil)

juncao = junta_dataframes(juncao, base_mortalidade_infantil, chave=['country', 'year'], tipo_juncao='left')
juncao = juncao.drop(columns=['childMortalityPer1000Born_old'])
juncao.describe()

Dataframes unidos com sucesso! 
Dimensão: (1704, 18)


year      lifeExp      gdpPercap  oneSidedViolence  \
count  1704.00000  1704.000000    1704.000000        520.000000   
mean   1979.50000    59.475235    7215.327095       1843.396154   
std      17.26533    12.916468    9857.454534      34328.028878   
min    1952.00000    23.600000     241.170000          0.000000   
25%    1965.75000    48.200000    1202.060000          0.000000   
50%    1979.50000    60.750000    3531.845000          0.000000   
75%    1993.25000    70.800000    9325.462500         46.500000   
max    2007.00000    82.600000  113523.130000     782261.000000   

       nonStateConflict  extrasystemic    intrastate    interstate  \
count        520.000000          520.0    520.000000    520.000000   
mean         152.476923            0.0   1065.101923    256.819231   
std          569.971029            0.0   4053.362937   4322.069164   
min            0.000000            0.0      0.000000      0.000000   
25%            0.000000            0.0      0.000000      0.000000   
50%            0.000000            0.0      0.000000      0.000000   
75%            0.000000            0.0    100.500000      0.000000   
max         5817.000000            0.0  43871.000000  95934.000000   

       totalWarDeaths  childMortalityPer1000Born_old  popDensityPerSquareKm  \
count      520.000000                    1521.000000            1548.000000   
mean      3317.794231                     116.850302             112.672720   
std      35009.273292                     100.116115             379.720921   
min          0.000000                       2.850000               0.520000   
25%          0.000000                      28.350000              14.285000   
50%          0.000000                      87.970000              42.765000   
75%        372.750000                     185.910000             104.460000   
max     785697.000000                     451.010000            6729.890000   

       humanDevelopmentIndex  gdpPerCaptaInflactionAdjusted  \
count             468.000000                    1656.000000   
mean                0.633141                    9534.027319   
std                 0.177727                   11681.196053   
min                 0.220000                     433.690000   
25%                 0.480000                    1913.582500   
50%                 0.660000                    4581.085000   
75%                 0.772500                   12199.975000   
max                 0.940000                   88925.070000   

       extremePovertyRate  childMortalityPer1000Born  
count         1589.000000                1629.000000  
mean            29.899880                 116.648115  
std             29.521719                 100.296925  
min              0.010000                   2.850000  
25%              1.440000                  28.360000  
50%             19.060000                  88.140000  
75%             54.910000                 185.310000  
max             99.300000                 458.930000

Agora ao item 8:

In [ ]:
renomeia_coluna(juncao, nome_anterior='popDensityPerSquareKm', novo_nome='popDensityPerSquareKm_old', inplace=True)

base_densidade_populacional = corrige_paises_estranhos(base_densidade_populacional)

juncao = junta_dataframes(juncao, base_densidade_populacional, chave=['country', 'year'], tipo_juncao='left')
juncao = juncao.drop(columns=['popDensityPerSquareKm_old'])
juncao.describe()

Dataframes unidos com sucesso! 
Dimensão: (1704, 19)


year      lifeExp      gdpPercap  oneSidedViolence  \
count  1704.00000  1704.000000    1704.000000        520.000000   
mean   1979.50000    59.475235    7215.327095       1843.396154   
std      17.26533    12.916468    9857.454534      34328.028878   
min    1952.00000    23.600000     241.170000          0.000000   
25%    1965.75000    48.200000    1202.060000          0.000000   
50%    1979.50000    60.750000    3531.845000          0.000000   
75%    1993.25000    70.800000    9325.462500         46.500000   
max    2007.00000    82.600000  113523.130000     782261.000000   

       nonStateConflict  extrasystemic    intrastate    interstate  \
count        520.000000          520.0    520.000000    520.000000   
mean         152.476923            0.0   1065.101923    256.819231   
std          569.971029            0.0   4053.362937   4322.069164   
min            0.000000            0.0      0.000000      0.000000   
25%            0.000000            0.0      0.000000      0.000000   
50%            0.000000            0.0      0.000000      0.000000   
75%            0.000000            0.0    100.500000      0.000000   
max         5817.000000            0.0  43871.000000  95934.000000   

       totalWarDeaths  childMortalityPer1000Born_old  \
count      520.000000                    1521.000000   
mean      3317.794231                     116.850302   
std      35009.273292                     100.116115   
min          0.000000                       2.850000   
25%          0.000000                      28.350000   
50%          0.000000                      87.970000   
75%        372.750000                     185.910000   
max     785697.000000                     451.010000   

       popDensityPerSquareKm_old  humanDevelopmentIndex  \
count                1548.000000             468.000000   
mean                  112.672720               0.633141   
std                   379.720921               0.177727   
min                     0.520000               0.220000   
25%                    14.285000               0.480000   
50%                    42.765000               0.660000   
75%                   104.460000               0.772500   
max                  6729.890000               0.940000   

       gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
count                    1656.000000         1589.000000   
mean                     9534.027319           29.899880   
std                     11681.196053           29.521719   
min                       433.690000            0.010000   
25%                      1913.582500            1.440000   
50%                      4581.085000           19.060000   
75%                     12199.975000           54.910000   
max                     88925.070000           99.300000   

       childMortalityPer1000Born  popDensityPerSquareKm  
count                1629.000000            1656.000000  
mean                  116.648115             113.762923  
std                   100.296925             369.073832  
min                     2.850000               0.520000  
25%                    28.360000              14.320000  
50%                    88.140000              42.765000  
75%                   185.310000             109.402500  
max                   458.930000            6729.890000

Por fim, o item 9:

In [45]:
hdinulo = juncao[juncao['humanDevelopmentIndex'].isnull()].groupby('country').size().reset_index(name='Quantidade').sort_values(by='Quantidade', ascending=False)
hdinulo.head(100)

country  Quantidade
105        Reunion          12
134  United States          12
121      Swaziland          12
120          Sudan          12
125         Taiwan          12
..             ...         ...
92       Nicaragua           8
91     New Zealand           8
90     Netherlands           8
89           Nepal           8
94         Nigeria           8

[100 rows x 2 columns]

Essa aqui o assunto é mais delicado. Ela tem dados da maior parte dos países de forma parcial, o que torna bem difícil da tratativa da tabela. Vamos corrigir o nome dos países que estão errados que conhecemos, mas isso não resolverá o caso dos países que só possuem dados parciais pro intervalo da tabela principal.

In [ ]:
renomeia_coluna(juncao, nome_anterior='humanDevelopmentIndex', novo_nome='humanDevelopmentIndex_old', inplace=True)

base_hdi = corrige_paises_estranhos(base_hdi)

juncao = junta_dataframes(juncao, base_hdi, chave=['country', 'year'], tipo_juncao='left')
juncao = juncao.drop(columns=['humanDevelopmentIndex_old'])
juncao.describe()

Dataframes unidos com sucesso! 
Dimensão: (1704, 20)


year      lifeExp      gdpPercap  oneSidedViolence  \
count  1704.00000  1704.000000    1704.000000        520.000000   
mean   1979.50000    59.475235    7215.327095       1843.396154   
std      17.26533    12.916468    9857.454534      34328.028878   
min    1952.00000    23.600000     241.170000          0.000000   
25%    1965.75000    48.200000    1202.060000          0.000000   
50%    1979.50000    60.750000    3531.845000          0.000000   
75%    1993.25000    70.800000    9325.462500         46.500000   
max    2007.00000    82.600000  113523.130000     782261.000000   

       nonStateConflict  extrasystemic    intrastate    interstate  \
count        520.000000          520.0    520.000000    520.000000   
mean         152.476923            0.0   1065.101923    256.819231   
std          569.971029            0.0   4053.362937   4322.069164   
min            0.000000            0.0      0.000000      0.000000   
25%            0.000000            0.0      0.000000      0.000000   
50%            0.000000            0.0      0.000000      0.000000   
75%            0.000000            0.0    100.500000      0.000000   
max         5817.000000            0.0  43871.000000  95934.000000   

       totalWarDeaths  childMortalityPer1000Born_old  \
count      520.000000                    1521.000000   
mean      3317.794231                     116.850302   
std      35009.273292                     100.116115   
min          0.000000                       2.850000   
25%          0.000000                      28.350000   
50%          0.000000                      87.970000   
75%        372.750000                     185.910000   
max     785697.000000                     451.010000   

       popDensityPerSquareKm_old  humanDevelopmentIndex_old  \
count                1548.000000                 468.000000   
mean                  112.672720                   0.633141   
std                   379.720921                   0.177727   
min                     0.520000                   0.220000   
25%                    14.285000                   0.480000   
50%                    42.765000                   0.660000   
75%                   104.460000                   0.772500   
max                  6729.890000                   0.940000   

       gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
count                    1656.000000         1589.000000   
mean                     9534.027319           29.899880   
std                     11681.196053           29.521719   
min                       433.690000            0.010000   
25%                      1913.582500            1.440000   
50%                      4581.085000           19.060000   
75%                     12199.975000           54.910000   
max                     88925.070000           99.300000   

       childMortalityPer1000Born  popDensityPerSquareKm  humanDevelopmentIndex  
count                1629.000000            1656.000000             497.000000  
mean                  116.648115             113.762923               0.633602  
std                   100.296925             369.073832               0.178918  
min                     2.850000               0.520000               0.220000  
25%                    28.360000              14.320000               0.480000  
50%                    88.140000              42.765000               0.650000  
75%                   185.310000             109.402500               0.780000  
max                   458.930000            6729.890000               0.940000

Mesmo após o tratamento, a completude dos dados desses países ficou muito aquém do esperado, então removemos essa tabela de forma definitiva da junção, pois sua presença provavelmente mais atrapalharia do que ajudaria.

In [48]:

juncao = juncao.drop(columns=['humanDevelopmentIndex'])
juncao.describe()

year      lifeExp      gdpPercap  oneSidedViolence  \
count  1704.00000  1704.000000    1704.000000        520.000000   
mean   1979.50000    59.475235    7215.327095       1843.396154   
std      17.26533    12.916468    9857.454534      34328.028878   
min    1952.00000    23.600000     241.170000          0.000000   
25%    1965.75000    48.200000    1202.060000          0.000000   
50%    1979.50000    60.750000    3531.845000          0.000000   
75%    1993.25000    70.800000    9325.462500         46.500000   
max    2007.00000    82.600000  113523.130000     782261.000000   

       nonStateConflict  extrasystemic    intrastate    interstate  \
count        520.000000          520.0    520.000000    520.000000   
mean         152.476923            0.0   1065.101923    256.819231   
std          569.971029            0.0   4053.362937   4322.069164   
min            0.000000            0.0      0.000000      0.000000   
25%            0.000000            0.0      0.000000      0.000000   
50%            0.000000            0.0      0.000000      0.000000   
75%            0.000000            0.0    100.500000      0.000000   
max         5817.000000            0.0  43871.000000  95934.000000   

       totalWarDeaths  childMortalityPer1000Born_old  \
count      520.000000                    1521.000000   
mean      3317.794231                     116.850302   
std      35009.273292                     100.116115   
min          0.000000                       2.850000   
25%          0.000000                      28.350000   
50%          0.000000                      87.970000   
75%        372.750000                     185.910000   
max     785697.000000                     451.010000   

       popDensityPerSquareKm_old  humanDevelopmentIndex_old  \
count                1548.000000                 468.000000   
mean                  112.672720                   0.633141   
std                   379.720921                   0.177727   
min                     0.520000                   0.220000   
25%                    14.285000                   0.480000   
50%                    42.765000                   0.660000   
75%                   104.460000                   0.772500   
max                  6729.890000                   0.940000   

       gdpPerCaptaInflactionAdjusted  extremePovertyRate  \
count                    1656.000000         1589.000000   
mean                     9534.027319           29.899880   
std                     11681.196053           29.521719   
min                       433.690000            0.010000   
25%                      1913.582500            1.440000   
50%                      4581.085000           19.060000   
75%                     12199.975000           54.910000   
max                     88925.070000           99.300000   

       childMortalityPer1000Born  popDensityPerSquareKm  
count                1629.000000            1656.000000  
mean                  116.648115             113.762923  
std                   100.296925             369.073832  
min                     2.850000               0.520000  
25%                    28.360000              14.320000  
50%                    88.140000              42.765000  
75%                   185.310000             109.402500  
max                   458.930000            6729.890000

Por fim, escreveremos todos os arquivos na pasta de bases tratadas, pra futuras verificações individuais e junções, caso necessário. (lembrando que os arquivos originais tem mais dados de cada país do que o arquivo principal, mas não levaremos isso em conta de forma direta neste trabalho, apenas os usaremos através do arquivo já agregado)

In [49]:
#Escreve o arquivo em uma base tratada
exporta_csv(base_limpa, '../../avaliacao-python-ciencia-dados/Bases/tratada/gapminder_tratada.csv', separador=';', encoding='utf-8', index=False)
exporta_csv(mortes_guerra, '../../avaliacao-python-ciencia-dados/Bases/tratada/mortes_guerra_tratada.csv', separador=';', encoding='utf-8', index=False)
exporta_csv(base_gdp_ajustado, '../../avaliacao-python-ciencia-dados/Bases/tratada/gdp_ajustado_inflacao_tratada.csv', separador=';', encoding='utf-8', index=False)
exporta_csv(base_extrema_pobreza, '../../avaliacao-python-ciencia-dados/Bases/tratada/base_extrema_pobreza_tratada.csv', separador=';', encoding='utf-8', index=False)
exporta_csv(base_mortalidade_infantil, '../../avaliacao-python-ciencia-dados/Bases/tratada/base_mortalidade_infantil_tratada.csv', separador=';', encoding='utf-8', index=False)
exporta_csv(base_densidade_populacional, '../../avaliacao-python-ciencia-dados/Bases/tratada/base_densidade_populacional_tratada.csv', separador=';', encoding='utf-8', index=False)
exporta_csv(base_hdi, '../../avaliacao-python-ciencia-dados/Bases/tratada/base_hdi_tratada.csv', separador=';', encoding='utf-8', index=False)
exporta_csv(juncao, '../../avaliacao-python-ciencia-dados/Bases/tratada/gapminder_juncao_tratada.csv', separador=';', encoding='utf-8', index=False)


Exportando o dataframe para o arquivo: ../../avaliacao-python-ciencia-dados/Bases/tratada/gapminder_tratada.csv ...
Dataframe exportado com sucesso!
Exportando o dataframe para o arquivo: ../../avaliacao-python-ciencia-dados/Bases/tratada/mortes_guerra_tratada.csv ...
Dataframe exportado com sucesso!
Exportando o dataframe para o arquivo: ../../avaliacao-python-ciencia-dados/Bases/tratada/gdp_ajustado_inflacao_tratada.csv ...
Dataframe exportado com sucesso!
Exportando o dataframe para o arquivo: ../../avaliacao-python-ciencia-dados/Bases/tratada/base_extrema_pobreza_tratada.csv ...
Dataframe exportado com sucesso!
Exportando o dataframe para o arquivo: ../../avaliacao-python-ciencia-dados/Bases/tratada/base_mortalidade_infantil_tratada.csv ...
Dataframe exportado com sucesso!
Exportando o dataframe para o arquivo: ../../avaliacao-python-ciencia-dados/Bases/tratada/base_densidade_populacional_tratada.csv ...
Dataframe exportado com sucesso!
Exportando o dataframe para o arquivo: ../../

---
## 3. Funções de Visualização e Análise Gráfica

### Objetivo
Criar funções reutilizáveis para gerar gráficos e apoiar a análise.

### Itens a incluir
- Definição de **funções dinâmicas** para gráficos (usando parâmetros flexíveis).  
- Tipos de gráficos recomendados:
  - Evolução temporal (ex: expectativa de vida ao longo dos anos).  
  - Comparações entre continentes.  
  - Relações entre PIB per capita, expectativa de vida e população.  
- Exemplos de uso das funções com resultados visuais.  

### Markdown deve conter
- Explicação do propósito de cada função.  
- Descrição dos tipos de visualizações utilizadas.  
- Observações sobre padrões ou correlações identificadas.

In [ ]:
# Funcao para elaborar grafico Line Plot
def plot_line( data, x, y, titulo, label_x, label_y, hue=None, marker=None, color=None, figsize=(12, 5), fontsize=16, loc_legenda='upper left'):

    plt.figure(figsize=figsize)
    sns.lineplot(data=data, x=x, y=y, hue=hue, marker=marker, color=color)
    plt.title(titulo, fontsize=fontsize)
    plt.xlabel(label_x)
    plt.ylabel(label_y)
    
    if hue:
        plt.legend(title=hue.capitalize(), loc=loc_legenda)
    
    plt.grid(True, linestyle='--', alpha=0.7)
    plt.tight_layout()
    plt.show()



# Funcao para elaborar grafico Bar Plot
def plot_bar( data, x, y, titulo, label_x, label_y, hue=None, palette='viridis', formato_rotulo=None, figsize=(12, 6), fontsize=16 ):

    plt.figure(figsize=figsize)
    sns.barplot(data=data, x=x, y=y, hue=hue, palette=palette)
    plt.xlabel(label_x)
    plt.ylabel(label_y)
    plt.title(titulo, fontsize=fontsize)
    plt.grid(True, linestyle='--', alpha=0.7, axis='y')
    plt.tight_layout()
    plt.show()



# Funcao para elaborar grafico Count Plot
def plot_count( data, x, titulo, label_x, label_y, palette='viridis', ordenar=True, figsize=(12, 6), fontsize=16 ):

    plt.figure(figsize=figsize)
    
    order = None
    if ordenar:
        order = data[x].value_counts().index
    
    sns.countplot(data=data, x=x, palette=palette, order=order)
    plt.xlabel(label_x)
    plt.ylabel(label_y)
    
    plt.title(titulo, fontsize=fontsize)
    plt.grid(True, linestyle='--')
    plt.tight_layout()
    plt.show()



# Funcao para elaborar grafico Heat Map
def plot_heat( data, titulo, cmap='viridis', annot=True, fmt=".2f", cbar_label=None, xlabel=None, ylabel=None, linewidths=0, figsize=(12, 6), fontsize=16 ):

    plt.figure(figsize=figsize)
    
    sns.heatmap( data=data, annot=annot, cmap=cmap, fmt=fmt, linewidths=linewidths, cbar_kws={'label': cbar_label} if cbar_label else None )
    
    plt.title(titulo, fontsize=fontsize)
    plt.xlabel(xlabel)
    plt.ylabel(ylabel)
    plt.yticks(rotation=0)
    plt.tight_layout()
    plt.show()



# Funcao para elaborar grafico Scatter Plot
def plot_scatter( data, x, y, titulo, label_x, label_y, hue=None, size=None, style=None, palette='viridis', tamanho_ponto=100, legenda_titulo=None, loc_legenda='best', figsize=(12, 7), fontsize=16) :

    plt.figure(figsize=figsize)
    
    sns.scatterplot( data=data, x=x, y=y, hue=hue, size=size, style=style, palette=palette, s=tamanho_ponto )
    
    plt.title(titulo, fontsize=fontsize)
    plt.xlabel(label_x)
    plt.ylabel(label_y)
    
    plt.grid(True, linestyle='--')
    plt.tight_layout()
    plt.show()

---
## 4. Interpretação das Visualizações

### Objetivo
Interpretar criticamente os dados com base nos gráficos e tabelas.

### Itens a incluir
- Explicação de cada gráfico gerado.  
- Justificativa da escolha do tipo de visualização.  
- Interpretação dos resultados observados.  
- Discussão sobre **tendências globais** e **diferenças regionais**.

---
## 5. Síntese dos Principais Insights

### Objetivo
Resumir os principais achados da análise.

### Itens a incluir
- **Padrões e tendências principais** (ex: crescimento da expectativa de vida, desigualdade econômica).  
- **Anomalias** ou exceções notáveis.  
- **Correlação** entre os indicadores analisados (PIB, população, expectativa de vida).  

### Markdown deve conter
- **Resumo analítico** dos resultados obtidos.  
- **Aplicações práticas**:
  - Implicações para políticas públicas.  
  - Relação com o desenvolvimento social e econômico.  
  - Observações sobre desigualdade global.

---
## 6. Conclusão e Reflexão

### Objetivo
Refletir sobre o processo de análise e o aprendizado obtido.

### Itens a incluir
- **Maiores desafios enfrentados** durante o trabalho.  
- **Principais aprendizados** obtidos.  
- **Habilidades desenvolvidas** (técnicas e analíticas).  
- **Reflexão final** sobre o impacto e relevância da análise.

#### Membros
1. Andre Augusto Ferreira Cabelo
2. Bruno Braga Guedes Cardoso
3. Bruno dos Santos Crivelli
4. Marcelo Tranche de Souza Junior